### ISSUE TITLE
How to calculate FLOPs with/without CSP-ization layer?

### ISSUE THREAD

**Issue by onelife90** [WongKinYiu/ScaledYOLOv4#358]
While reading the Scaled-YOLov4 paper, the calculation of various layers with different model scaling coefficients is not understood. Is there anyone who can explain the calculation of FLOPs of the darknet layer in table 2?

![image](https://user-images.githubusercontent.com/64455878/139629659-c4a94f10-1772-4ce4-b8b4-c0f725441151.png)

Also, please tell me if my understanding of the other two model layers is correct.
First, this is my understanding of ResNet layer.

![image](https://user-images.githubusercontent.com/64455878/139629707-1cecdd5a-6e20-40c9-91be-4e955892b90d.png)

17HWC^2 = 4HWC^2 + 9HWC^2 + 4HWC^2

Looking at total FLOPs (17HWC^2) in the paper, it was assumed that the residual block was explained and understood.

![image](https://user-images.githubusercontent.com/64455878/139630066-f4f4cc09-bf0f-4c5e-9f33-aa11a015a75c.png)

/16 is divided because there are 16 residual blocks of ResNet34-layer.

Second, it is the FLOPs calculation for the ResNeXt layer.

![image](https://user-images.githubusercontent.com/64455878/139629909-5c260fce-edc7-435f-a3ce-0798f0873a64.png)

![image](https://user-images.githubusercontent.com/64455878/139630345-2e01fc5b-11d4-44c4-b3fb-db8beaef5b9a.png)

137HWC^2 = 64HWC^2 + 9HWC^2 + 64HWC^2

![image](https://user-images.githubusercontent.com/64455878/139630620-5ada1032-a7ae-4f00-9dea-f5bba199e61f.png)

Finally, the reason for dividing 128 in the FLOPs calculation of the ResNeXt layer is that when divided into 32 paths, the grouped convolution has 128 widths (#channels).

If it is the FLOPs calculation method I understood, the FLOPs calculation for DarkNet layer should be 20WHC^2, which is marked as 5WHC^2 in the paper.
Isn't it right that the residual block on Darknet is shown in the figure?
My head hurts so much. Help me.

![image](https://user-images.githubusercontent.com/64455878/139630965-4190e344-18a7-4789-916a-c2ea3d6e5dce.png)

And if you are more kind and know the calculation process of CSP layer well, please explain.

**Comment by WongKinYiu** [comment]
> /16 is divided because there are 16 residual blocks of ResNet34-layer.

No, I assume channel of base layer is `b` in the table. So all of `C`s in your figure will become  `C/4,` and `/16` come from `(C/4)*(C/4)`. Number of residual block is `k` in the table. So do resx, dark, and csp models.

**Comment by onelife90** [comment]
Thank for your reply.
If I understood correctly, Darknet will be `/4` because it is `(C/2)*(C/2)`, but in table 2, it will be `20WHKb^2/4=5WHKb^2`, please check if it is correct.
If not, please explain it one more time.

**Comment by WongKinYiu** [comment]
> As you explained, ResNeXt will be `/128` because it is `(C/64)*(C/64)`.

No, `(C/4)*(C/4)` leads `/16`, `(C/64)*(C/64)` leads `/4096`.
